src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java

```diff
@@ -0,0 +1,110 @@
+ */
+package org.candlepin.subscriptions.subscription;
+
+import org.candlepin.subscriptions.db.SubscriptionRepository;
+import org.candlepin.subscriptions.subscription.api.model.Subscription;
+
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Optional;
+
+/**
+ * Update subscriptions from subscription service responses.
+ */
+@Component
+public class SubscriptionSyncController {
+
+    private final SubscriptionRepository subscriptionRepository;
+
+    public SubscriptionSyncController(SubscriptionRepository subscriptionRepository) {
+        this.subscriptionRepository = subscriptionRepository;
+    }
+
+    public void syncSubscription(Subscription subscription) {
+        final Optional<org.candlepin.subscriptions.db.model.Subscription> entity =
+            subscriptionRepository.findActiveSubscription(String.valueOf(subscription.getId()));
+        if (entity.isPresent()) {
+            final org.candlepin.subscriptions.db.model.Subscription dbSub = entity.get();
+            if (updateSubscription(subscription, dbSub)) {
+                // end the current record
+                dbSub.setEndDate(OffsetDateTime.now());
+                subscriptionRepository.save(dbSub);
+                // create a new record
+                final org.candlepin.subscriptions.db.model.Subscription newSub =
+                    new org.candlepin.subscriptions.db.model.Subscription();
+                newSub.setSubscriptionId(dbSub.getSubscriptionId());
+                newSub.setSku(dbSub.getSku());
+                newSub.setOwnerId(dbSub.getOwnerId());
+                newSub.setQuantity(subscription.getQuantity());
+                newSub.setStartDate(OffsetDateTime.now());
+                if (subscription.getEffectiveEndDate() != null) {
+                    newSub.setEndDate(convertLongTime(subscription.getEffectiveEndDate()));
+                }
+                subscriptionRepository.save(newSub);
+            }
+            else {
+                subscriptionRepository.save(dbSub);
+            }
+        }
+        else {
+            // create a new record
+            final org.candlepin.subscriptions.db.model.Subscription newSub =
+                new org.candlepin.subscriptions.db.model.Subscription();
+            newSub.setSubscriptionId(String.valueOf(subscription.getId()));
+            newSub.setQuantity(subscription.getQuantity());
+            if (subscription.getEffectiveStartDate() != null) {
+                newSub.setStartDate(convertLongTime(subscription.getEffectiveStartDate()));
+            }
+            if (subscription.getEffectiveEndDate() != null) {
+                newSub.setEndDate(convertLongTime(subscription.getEffectiveEndDate()));
+            }
+            subscriptionRepository.save(newSub);
+        }
+    }
+
+    private static boolean updateSubscription(Subscription dto,
```

Comment:
I'd like to see this get renamed/broken apart or something.  It's unclear what the method is meant to do, as it's name indicates it's updating a subscription - but it's returning a boolean.  Line 90 is returning true without updating the subscription entity at all.  The next few lines update the entity, but then return a false.  

It looks like it's intended to be a "does the subscription need to be updated" kind of flag...while having a side effect of updating the entity class anyway prior to it being saved on line 68.

Do these two private methods need to be static?

Reply:
No issue with splitting them up though I'd prefer to keep them static as they are not bound to the instance state and java has significant performance differences between static and member calls.